frontend/server/k8s-helper.ts

```diff
@@ -62,6 +62,62 @@ function getNameOfViewerResource(logdir: string): string {
   return 'viewer-' + crypto.SHA1(logdir);
 }
 
+function parseTensorboardLogDir(logdir: string, podTemplateSpec: object): string {
+  const urls: string[] = [];
+  const seriesParts = logdir.split(',');
+  for (const seriesPart of seriesParts) {
+    const strPath = seriesPart.replace(/Series\d*:/g, '');
+
+    // volume storage: parse real local mount path
+    if (strPath.startsWith('volume://')) {
+      const pathParts = strPath.substr('volume://'.length).split('/');
+      const bucket = pathParts[0];
+      const key = pathParts.slice(1).join('/');
+      // check volume exist
+      const volumes = podTemplateSpec['spec'] && podTemplateSpec['spec']['volumes'];
+      if (!Array.isArray(volumes) || !volumes.find(v => v?.name === bucket)) {
+        throw new Error(`Volume ${bucket} not configured`);
+      }
+
+      // get volume mount object, backend viewer CRD only use containers[0] as viewer main container
+      const volumeMounts =
+        podTemplateSpec['spec'] &&
+        podTemplateSpec['spec']['containers'] &&
+        podTemplateSpec['spec']['containers'][0] &&
+        podTemplateSpec['spec']['containers'][0]['volumeMounts'];
+
+      if (!Array.isArray(volumeMounts)) {
+        throw new Error(`Volume ${bucket} not mounted`);
+      }
+
+      // find volumes mount
+      const volumeMount = volumeMounts.find(v => {
+        // volume name must be same
+        if (v?.name !== bucket) {
+          return false;
+        }
+        if (v?.subPath) {
+          return key.startsWith(v.subPath);
+        }
+        return true;
+      });
+
+      if (!volumeMount) {
+        throw new Error(
+          `Volume ${bucket} not mounted or volume ${bucket} with subPath(which is prefix of ${key}) not mounted`,
+        );
+      }
+
+      // finally file path
+      const filePath = pruneAndConcatPath(volumeMount.mountPath, key, volumeMount.subPath);
```

Comment:
nit: Can we extract this as a function?

We've spent a lot of efforts discussing desired behavior of this part... I think it's worth its own function and some documentation about how we are parsing volume here.

Reply:
actually, you've been duplicating this complex function twice, once in `artifact.ts` and once here.
Suggest extracting the function and reuse it in both places.

Because we already have well written error messages in `artifact.ts`, so I think it makes more sense to extract code there into a function that returns both result and error message, so it can easily be reused in both places.

Reply:
I extract this as a function into utils, any suggestion will be welcome.